File: sort/shell_sort/javascript/ShellSort.js
function shellSort (a) {
    for (var h = a.length; h > 0; h = parseInt(h / 2)) {
        for (var i = h; i < a.length; i++) {
            var k = a[i];
            for (var j = i; j >= h && k < a[j - h]; j -= h)
                a[j] = a[j - h];
            a[j] = k;
        }
    }
    return a;
}
var arrayToSort = [1,4,5,-4,2,56,345,23];
console.log('Not sorted array', arrayToSort);
shellSort(arrayToSort);
console.log('Sorted array', arrayToSort);

File: sort/shell_sort/ruby/ShellSort.rb
def shellsort(numbers)
  gap = numbers.length / 2
  #sort
  while gap > 0 do
    i = gap
    while i < numbers.length do
      temp = numbers[i]
      j = i
      while j >= gap &&  numbers[j-gap] > temp do
        numbers[j] = numbers[j-gap]
        j-=gap
      end
      numbers[j] = temp
      i+=1
    end
    	
    gap = gap / 2
  end
  return numbers
end

result =  shellsort(Array.new(10) { rand(1...1000) })
puts result.inspect

#test
pass = true
i = 0
while i < result.length-1 do
  if result[i] > result[i+1]
    pass = false
    break
  end
  i+=1
end

if pass
  puts "Success: Array is sorted."
else
  puts "FAIL: Array is not sorted."
end